### ISSUE TITLE
Xdebug is not working with Docker for Mac

### ISSUE THREAD

**Issue by achekulaev** [docksal/docksal#389]
Xdebug does not seem to be working with Docker for Mac.

**Comment by ApacheEx** [comment]
yeah, I had the same problem, as workaround I did:

1) add into `.docksal/etc/php/php.ini`:
```
[xdebug]
xdebug.remote_connect_back = 0
xdebug.idekey = "PHPSTORM"
xdebug.remote_host = 10.254.254.254
```

2) and create a new network alias on your host machine:
`sudo ifconfig lo0 alias 10.254.254.254`

some info here:
- https://gist.github.com/ralphschindler/535dc5916ccbd06f53c1b0ee5a868c93
- http://jamescowie.me/blog/2016/12/all-hail-xdebug-and-lets-let-var-dump-die

**Comment by lmakarov** [comment]
Try it with `xdebug.remote_host = 192.168.65.1`.

`192.168.65.0/24` is the network used by Docker for Mac by default. `192.168.65.1` can be used to access the host from within the Moby VM and containers.

Docksal is using `192.168.64.0/24`: `192.168.64.1` for host and `192.168.64.100` for the VM. On Linux the entire subnet is assigned as an alias on `lo0`.

You could technically change the default network Docker for Mac is using to `192.168.64.0/24`, but that will result in conflicts if you decide to switch back to Docker + VirtualBox or use both at the same time.

![image](https://user-images.githubusercontent.com/1205005/32914178-89309a2e-cac9-11e7-939f-e832ae8bf5a5.png)

**Comment by lmakarov** [comment]
# TL;DR

Short term, this can be used with Docker for Mac.

Add to `.docksal/etc/php/php.ini` and do `fin reset cli`:

```
[xdebug]
xdebug.remote_connect_back = 0
xdebug.remote_host = 192.168.65.1
```

In Docker for Mac, `192.168.65.1` is the host IP, which can be used by the Moby VM and containers to reach the host.

---

# The long story

[`xdebug.remote_connect_back`](https://xdebug.org/docs/all_settings#remote_connect_back) is the important part. It's set to `1` in our `cli` images by default, which tells xdebug to connect back for debugging to whomever initiated the original connection to the server. In this case, xdebug reads the remote server address from either `$_SERVER['HTTP_X_FORWARDED_FOR']` or `$_SERVER['REMOTE_ADDR']` variables.

On Docker for Mac, `docksal-vhost-proxy` sees host requests originating from the Docker bridge network interface, thus:

```
$_SERVER['HTTP_X_FORWARDED_FOR'] | 172.17.0.1
```

However, that bridge IP cannot be used to access the host back and the following will show up in xdebug logs ([xdebug.remote_log](https://xdebug.org/docs/all_settings#remote_log) to enabled logging):

```
$ fin exec tail -f /tmp/xdebug.log
Log opened at 2017-11-18 01:59:37
I: Checking remote connect back address.
I: Checking header 'HTTP_X_FORWARDED_FOR'.
I: Remote address found, connecting to 172.17.0.1:9000.
W: Creating socket for '172.17.0.1:9000', poll success, but error: Operation now in progress (29).
E: Could not connect to client. :-(
Log closed at 2017-11-18 01:59:37
```

With VirtualBox, host requests originate via the host-only network and `docksal-vhost-proxy` sees the correct host IP:

```
$_SERVER['HTTP_X_FORWARDED_FOR'] | 192.168.64.1
```

That one CAN be used by containers to reach back to the host.

With Docker for Mac/Windows, we've been setting the canonical Docksal IP (`192.168.64.100`), but have not been doing the same for the canonical host IP (`192.168.64.1`). They would both point to the host in this case (same as on Linux), but have to be distinguished to maintain the compatibility with the VirtualBox mode.

**Comment by achekulaev** [comment]
Needs to be added to docs/tools/xdebud.md

**Comment by lmakarov** [comment]
Yes, I'm updating the xdebug docs as well. Not finished yet.

**Comment by wireblue** [comment]
In case this helps anyone else, Docker 18.03 (and later) enable you to use a special hostname to connect to the host.

```
xdebug.remote_host = "host.docker.internal"
```

This means you don't need to get/use the host's IP address anymore.

NOTE: This is for development purposes only and will not work in a production environment outside of Docker for Mac.

**Comment by achekulaev** [comment]
@lmakarov should we update stacks with the above?

**Comment by ssuhss** [comment]
@wireblue Tnks, you save my life with this line!

**Comment by achekulaev** [comment]
@lmakarov looks like it is still the issue

<img width="554" alt="still the issue" src="https://user-images.githubusercontent.com/402007/60522820-b15b2700-9cb7-11e9-8d90-77afd297df48.png">

**Comment by JDDoesDev** [comment]
I'll update the xdebug docs if you would like?

**Comment by achekulaev** [comment]
Its not about the docs per se. If `xdebug.remote_host = "host.docker.internal"` works everywhere we should just update default stack files to use that and forget about the problem. 

But this needs to be tested across different OSs and different Docker setups (Vbox/Docker Desktop)

**Comment by lmakarov** [comment]
`host.docker.internal` works on Windows and Mac, but not on Linux.

Linux support for this is _almost_ there: https://github.com/docker/for-linux/issues/264

**Comment by lmakarov** [comment]
I've just double checked this. We do not need to rely on `host.docker.internal` at all, since we have our own canonical and persistent host IP address that works in all environments: `192.168.64.1`.

The default Docksal stack(s) ship with the necessary xdebug settings:
https://github.com/docksal/docksal/blob/c0f2a12aff4510677c764a0e69c06a9e368e6610/stacks/services.yml#L146-L149

I checked this with Docker Desktop for Mac and xdebug works fine for me using the docksal/boilerplate-drupal8 repo (`fin project create` => Drupal 8).

If xdebug does not work in a project, then the project stack us likely custom and outdated.
See if you can get it to work in a boilerplate setup (`fin project create`). If not, then file a new issue providing all necessary diagnostics information.
